From c33671a9bc1be9683549e40fea088b5e04ba36d9 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:22:13 -0400
Subject: [infiniband] ocrdma: Obtain SL from device structure

Message-id: <bdd05f79ef942818c7488ef48fa97f616d65a614.1411571481.git.dledford@redhat.com>
Patchwork-id: 96617
O-Subject: [Patch RHEL7 38/48] RDMA/ocrdma: Obtain SL from device structure
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Devesh Sharma <devesh.sharma@emulex.com>

Currently, driver obtains service level value from ah_attr->sl field.
However, this field is set to zero all the times from rdma-cm.  This
patch allows create_ah to obtain service level from dev->sl.

Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 0ea8726250cae3defa0c1065801017f26a269c93)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_ah.c b/drivers/infiniband/hw/ocrdma/ocrdma_ah.c
index a023234..40f8536 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_ah.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_ah.c
@@ -35,6 +35,8 @@
 #include "ocrdma_ah.h"
 #include "ocrdma_hw.h"
 
+#define OCRDMA_VID_PCP_SHIFT 0xD
+
 static inline int set_av_attr(struct ocrdma_dev *dev, struct ocrdma_ah *ah,
     struct ib_ah_attr *attr, int pdid)
 {
@@ -55,7 +57,7 @@ static inline int set_av_attr(struct ocrdma_dev *dev, struct ocrdma_ah *ah,
  if (vlan_tag && (vlan_tag < 0x1000)) {
   eth.eth_type = cpu_to_be16(0x8100);
   eth.roce_eth_type = cpu_to_be16(OCRDMA_ROCE_ETH_TYPE);
-  vlan_tag |= (attr->sl & 7) << 13;
+  vlan_tag |= (dev->sl & 0x07) << OCRDMA_VID_PCP_SHIFT;
   eth.vlan_tag = cpu_to_be16(vlan_tag);
   eth_sz = sizeof(struct ocrdma_eth_vlan);
   vlan_enabled = true;
diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_hw.c b/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
index a4d27c7..df68782 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_hw.c
@@ -2314,10 +2314,8 @@ static int ocrdma_set_av_params(struct ocrdma_qp *qp,
   cmd->params.vlan_dmac_b4_to_b5 |=
       vlan_id << OCRDMA_QP_PARAMS_VLAN_SHIFT;
   cmd->flags |= OCRDMA_QP_PARA_VLAN_EN_VALID;
-  /* override the sl with default priority if 0 */
   cmd->params.rnt_rc_sl_fl |=
-   (ah_attr->sl ? ah_attr->sl :
-    qp->dev->sl) << OCRDMA_QP_PARAMS_SL_SHIFT;
+   (qp->dev->sl & 0x07) << OCRDMA_QP_PARAMS_SL_SHIFT;
  }
  return 0;
 }
-- 
1.7.1